lib/abbreviato/truncated_sax_document.rb

```diff
@@ -0,0 +1,194 @@
+    end
+    append_to_truncated_string(string_to_append)
+  end
+
+  # This method is called when the parser encounters a comment
+  def comment(string)
+    comment = comment_tag(string)
+    if comment.bytesize <= remaining_length
+      append_to_truncated_string(comment)
+    else
+      @truncated = true
+    end
+  end
+
+  # This method is called when the parser encounters cdata. In practice, this also
+  # gets called for this style of comment inside an element:
+  #
+  #   <style><!--
+  #     /* Font Definitions */
+  #     @font-face
+  #       {font-family:Wingdings;
+  #       panose-1:5 0 0 0 0 0 0 0 0 0;}
+  #   --></style>
+  #
+  def cdata_block(string)
+    if string.bytesize <= remaining_length
+      append_to_truncated_string(string)
+    else
+      @truncated = true
+    end
+  end
+
+  # This method is called when the parser encounters a closing tag
+  def end_element(name)
+    if ignore_mode?
+      exit_ignored_level unless single_tag_element?(name)
+      return
+    end
+    return if max_length_reached || ignorable_tag?(name)
+
+    unless single_tag_element?(name)
+      @closing_tags.pop
+      # Don't count the length when closing a tag - it was accommodated when
+      # the tag was opened
+      append_to_truncated_string(closing_tag(name), 0)
+    end
+  end
+
+  def end_document
+    @closing_tags.reverse.each { |name| append_to_truncated_string(closing_tag(name)) }
+  end
+
+  private
+
+  def opening_tag(name, attributes)
+    attributes_string = attributes_to_string(attributes)
+    if single_tag_element? name
+      "<#{name}#{attributes_string}/>"
+    else
+      "<#{name}#{attributes_string}>"
+    end
+  end
+
+  def comment_tag(comment)
+    "<!--#{comment}-->"
+  end
+
+  def closing_tag(name)
+    "</#{name}>"
+  end
+
+  def remaining_length
+    max_length - @estimated_length
+  end
+
+  def single_tag_element?(name)
+    VOID_TAGS.include? name
+  end
+
+  def append_to_truncated_string(string, overridden_length = nil)
+    @truncated_string << string
+    @estimated_length += (overridden_length || string.bytesize)
+  end
+
+  def attributes_to_string(attributes)
+    attributes.inject(' ') do |string, attribute|
+      key, value = attribute
+      string << "#{key}='#{@html_coder.encode value}' "
+    end.rstrip
+  end
+
+  def max_length_reached
+    @estimated_length >= max_length
+  end
+
+  def truncate_string(string)
+    @truncated = true
+    truncate_length = remaining_length - tail.bytesize
+    (string.byteslice(0, truncate_length) || '').scrub('')
+  end
+
+  def overridden_tag_length(tag_name, rendered_tag_with_attributes)
+    # Start with the opening tag
+    length = rendered_tag_with_attributes.bytesize
+
+    # Add on closing tag if necessary
+    length += closing_tag(tag_name).bytesize unless single_tag_element?(tag_name)
+    length
+  end
+
+  def ignorable_tag?(name)
+    IGNORABLE_TAGS.include?(name.downcase)
+  end
+
+  def enter_ignored_level
+    @truncated = true
```

Comment:
seems like we need to move the checks on `single_tag_element?` that accompany the calls to this function to move into the function itself so that we still get the `@truncated = true` for the single tag elements